### PR TITLE
Add success and failure callbacks for plan builds

### LINKED ISSUE #5054: Add success/failure callbacks to the package build process
It can be useful to have callbacks that fire at the end of a package build, whether successful or a failure, in order to help integrate the Habitat build process into a user's existing software development pipeline.

### PR DESCRIPTION
**Pull request by christophermaier**
Does what it says on the tin; check out commit messages for further details.

Linux-only right now, pending me getting a Windows dev environment and learning PowerShell 😄 

Closes #5054 

cc: @bixu @smartb-pair

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

**Comment by bixu** [comment]
♥️

**Comment by fnichol** [comment]
Haay! I forgot about this in our effort to ship the last release. I'll aim to put this through its paces this week so we can get it in.

**Comment by reset** [comment]
Why wouldn't somebody just check if `hab build` succeeded or failed and then do something in a separate process? It would seem to me this addition would tie somebody's internal post-processing to the build itself. I have always envisioned the plan as a unit of build work and it's responsibilities stop there.

I would suggest that we don't let the Studio or Plans know about post-processing steps and leave that up to Builder or another build server's pipeline on a successful build of the Plan. I think coupling the building unit to a build server could result in either less portable building units or the need to disable features depending on the environment you're calling the build unit from.

**Comment by bixu** [comment]
Yes, we could do it elsewhere, but for me this breaks one of the most beautiful properties of a Plan - its self-contained nature. The more I have to train users to be experts in Builder, the more slowly Habitat gets real traction.

Also, at least a my org, we need post success/failure actions both during Builder builds and during local Studio builds - having to maintain (different) glue code in both environments would be a bummer.

**Comment by bixu** [comment]
Where are we in this discussion? We’ve run into a fair handful of cases in the past few weeks where `do_after_failure` would have saved us time and frustration (and where we needed the feature in both Builder and the local Studio).

**Comment by christophermaier** [comment]
Sorry, looks like this fell through the cracks with everything that's been
going on lately.

@fnichol any thoughts on this?

File: components/plan-build/bin/hab-plan-build.sh [hunk 1]
```diff
@@ -514,7 +512,34 @@ _on_exit() {
 # * TERM (15)
 # * ERR - when a shell command raises an error. Useful for `set -e; set -E`
 #   above.
-trap _on_exit 1 2 3 15 ERR
+# * EXIT - ensure we wrap things up properly (including firing any
+#   `do_after_failure` callback) if any thing calls `exit` (such as `exit_with`)
+#
+# See also `_do_final_callback_wrapper()`; it does some trap munging, too.
+trap _on_exit 1 2 3 15 ERR EXIT
```
Comment (fnichol) [inline components/plan-build/bin/hab-plan-build.sh]: Ah clever, nice approach!

File: components/plan-build/bin/hab-plan-build.sh [hunk 2]
```diff
@@ -2201,6 +2226,68 @@ do_default_end() {
...
+        # Thus, we need to do a lot of juggling of error handling.
+        #
+        # Kinda gross, but it's Bash.
+
+        # Set this to prevent an error from the hook from triggering
+        # the error handler.
+        set +e
+        # Don't want an error from the success hook to trigger the error hook
+        trap - ERR
+        (
```
Comment (fnichol) [inline components/plan-build/bin/hab-plan-build.sh]: One consequence of this approach is that any side effects of the `${callback_function}` would be in a sub-shell and would not affect the main process. That is, mutating any variables wouldn't affect the main program. Right now I can't think of a good reason to need this...so we're probably okay here. Is this intended?

File: components/plan-build/bin/hab-plan-build.sh [hunk 3]
```diff
@@ -2506,4 +2598,4 @@ build_line "Blake2b Checksum: $_pkg_blake2bsum"
 build_line
 build_line "I love it when a plan.sh comes together."
 build_line
-_on_exit  0
```
Comment (fnichol) [inline components/plan-build/bin/hab-plan-build.sh]: Oh spacing!

**Comment by fnichol** [comment]
Nice behavior even when I screw up (as a Plan author):

```sh
> cat failur/plan.sh 
pkg_origin=fnichol
pkg_name="failur"

do_after_failure() {
  build_line "Oh no, failur..."
  set > /tmp/failur.log
}

> hab studio enter
# ...

[1][default:/src:0]# /src/habitat/components/plan-build/bin/hab-plan-build.sh failur                                                                       
   : Loading /src/failur/plan.sh
   failur: Plan loaded
   failur: Validating plan metadata
ERROR: Failed to build. 'pkg_version' must be set or 'pkg_version()' function must be implemented and then invoking by calling 'update_pkg_version()'.     
   failur: Build time: 0m0s
   failur: Exiting on error
   failur: 'do_after_failure' callback function is defined; executing...
   failur: Oh no, failur...
[2][default:/src:1]# ls -l /tmp/failur.log
-rw-r--r-- 1 root root 68223 Jul 11 17:53 /tmp/failur.log
```

**Comment by fnichol** [comment]
The success case also appears where expected:

```sh
> cat failur/plan.sh 
pkg_origin=fnichol
pkg_name="failur"
pkg_version="1.0.0"

do_after_failure() {
  build_line "Oh no, failur..."
  set > /tmp/failur.log
}

do_after_success() {
  build_line "I have great success"
}

do_build() {
  build_line "Building nothing"
}

do_install() {
  build_line "A very small package indeed"
}

[6][default:/src:1]# /src/habitat/components/plan-build/bin/hab-plan-build.sh failur                                                                       
   : Loading /src/failur/plan.sh                                                                                                                           
   failur: Plan loaded
   failur: Validating plan metadata                
   failur: Using HAB_BIN=/hab/bin/hab for installs, signing, and hashing
   failur: hab-plan-build.sh setup
   failur: Writing pre_build file
   failur: Resolving build dependencies
   failur: Resolving run dependencies
   failur: Setting up environment
   failur: Populating runtime environment from dependencies
   failur: Populating buildtime environment from dependencies
   failur: Layering runtime environment on top of system environment
   failur: Layering buildtime environment on top of system environment
   failur: Setting PATH=/hab/bin:/hab/pkgs/core/hab-plan-build/0.58.0/20180629150501/bin:/hab/pkgs/core/diffutils/3.3/20170513214227/bin:/hab/pkgs/core/less
/481/20170513213936/bin:/hab/pkgs/core/make/4.2.1/20170513214620/bin:/hab/pkgs/core/mg/20160118/20170513220846/bin:/hab/pkgs/core/util-linux/2.27.1/20170513
214732/bin:/hab/pkgs/core/vim/8.0.0004/20170513220742/bin:/hab/pkgs/core/ncurses/6.0/20170513213009/bin:/hab/pkgs/core/acl/2.2.52/20170513213108/bin:/hab/pk
gs/core/attr/2.4.47/20170513213059/bin:/hab/pkgs/core/bash/4.3.42/20170513213519/bin:/hab/pkgs/core/binutils/2.25.1/20170513201927/bin:/hab/pkgs/core/bzip2/
1.0.6/20170513212938/bin:/hab/pkgs/core/coreutils/8.25/20170513213226/bin:/hab/pkgs/core/file/5.24/20170513201915/bin:/hab/pkgs/core/findutils/4.4.2/2017051
3214305/bin:/hab/pkgs/core/gawk/4.1.3/20170513213646/bin:/hab/pkgs/core/glibc/2.22/20170513201042/bin:/hab/pkgs/core/grep/2.22/20170513213444/bin:/hab/pkgs/
core/gzip/1.6/20170513214605/bin:/hab/pkgs/core/hab/0.58.0/20180629144346/bin:/hab/pkgs/core/libcap/2.24/20170513213120/bin:/hab/pkgs/core/libidn/1.32/20170
513215043/bin:/hab/pkgs/core/openssl/1.0.2l/20171014213633/bin:/hab/pkgs/core/pcre/8.38/20170513213423/bin:/hab/pkgs/core/rq/0.9.2/20170612005822/bin:/hab/p
kgs/core/sed/4.2.2/20170513213123/bin:/hab/pkgs/core/tar/1.29/20170513213607/bin:/hab/pkgs/core/unzip/6.0/20180310001951/bin:/hab/pkgs/core/wget/1.19.1/2017
1024102323/bin:/hab/pkgs/core/xz/5.2.2/20170513214327/bin
   failur: Clean the cache
   failur: Setting build environment
mkdir: created directory '/hab/cache/src/failur-1.0.0'
   failur: Setting PREFIX=/hab/pkgs/fnichol/failur/1.0.0/20180711180513
   failur: Setting LD_RUN_PATH=
   failur: Setting CFLAGS=
   failur: Setting CXXFLAGS=
   failur: Setting CPPFLAGS=
   failur: Setting LDFLAGS=
   failur: Setting PKG_CONFIG_PATH=
   failur: Preparing to build
   failur: Building
   failur: Building nothing
   failur: Installing
mkdir: created directory '/hab/pkgs/fnichol'
mkdir: created directory '/hab/pkgs/fnichol/failur'
mkdir: created directory '/hab/pkgs/fnichol/failur/1.0.0'
mkdir: created directory '/hab/pkgs/fnichol/failur/1.0.0/20180711180513'
   failur: A very small package indeed
   failur: Writing configuration
   failur: Writing service management scripts
   failur: Stripping unneeded symbols from binaries and libraries
   failur: Building package metadata
   failur: Generating blake2b hashes of all files in the package
   failur: Generating signed metadata FILES
» Signing failur_blake2bsums
☛ Signing failur_blake2bsums with fnichol-20161003171546 to create /hab/pkgs/fnichol/failur/1.0.0/20180711180513/FILES
★ Signed artifact /hab/pkgs/fnichol/failur/1.0.0/20180711180513/FILES.
   failur: Creating manifest
   failur: Generating package artifact
/hab/pkgs/core/tar/1.29/20170513213607/bin/tar: Removing leading `/' from member names
/hab/cache/artifacts/.fnichol-failur-1.0.0-20180711180513-x86_64-linux.tar (1/1)
  100 %           1164 B / 10.0 KiB = 0.114
» Signing /hab/cache/artifacts/.fnichol-failur-1.0.0-20180711180513-x86_64-linux.tar.xz
☛ Signing /hab/cache/artifacts/.fnichol-failur-1.0.0-20180711180513-x86_64-linux.tar.xz with fnichol-20161003171546 to create /hab/cache/artifacts/fnichol-f
ailur-1.0.0-20180711180513-x86_64-linux.hart
★ Signed artifact /hab/cache/artifacts/fnichol-failur-1.0.0-20180711180513-x86_64-linux.hart.
'/hab/cache/artifacts/fnichol-failur-1.0.0-20180711180513-x86_64-linux.hart' -> '/src/results/fnichol-failur-1.0.0-20180711180513-x86_64-linux.hart'
   failur: hab-plan-build.sh cleanup
   failur: 'do_after_success' callback function is defined; executing...
   failur: I have great success
   failur: 
   failur: Source Path: /src/failur
   failur: Installed Path: /hab/pkgs/fnichol/failur/1.0.0/20180711180513
   failur: Artifact: /src/results/fnichol-failur-1.0.0-20180711180513-x86_64-linux.hart
   failur: Build Report: /src/results/last_build.env
   failur: SHA256 Checksum: ac0956faa3b13d678a218d771471589d264124b8eed590828c33fd551d492d21
   failur: Blake2b Checksum: 84f2a7249ef1dc6e761ec48adc25cdfaf325df646ce06e4c416ba71e33b98de0
   failur: 
   failur: I love it when a plan.sh comes together.
   failur: 
   failur: Build time: 0m0s
```

**Comment by bixu** [comment]
♥️

**Comment by christophermaier** [comment]
tenor-171175631

**Pull request merged**